File: java17/src/main/java/folio/codinginterview/presentation/MarketPriceController.java
package folio.codinginterview.presentation;

import folio.codinginterview.application.usecase.market_price.UpdateMarketPriceUsecase;
import folio.codinginterview.domain.StockSymbol;
import folio.codinginterview.presentation.PresentationException.BadRequestException;

import java.math.BigDecimal;
import java.util.ArrayList;
import java.util.List;
import java.util.Optional;
import java.util.concurrent.CompletableFuture;

public final class MarketPriceController {
    public record MarketPriceItemDto(String symbol, String market_price) {}

    public record UpdateMarketPriceRequest(List<MarketPriceItemDto> market_prices) {}

    private final UpdateMarketPriceUsecase updateMarketPriceUsecase;

    public MarketPriceController(UpdateMarketPriceUsecase updateMarketPriceUsecase) {
        this.updateMarketPriceUsecase = updateMarketPriceUsecase;
    }

    public CompletableFuture<Void> updateMarketPrice(UpdateMarketPriceRequest req) {
        List<UpdateMarketPriceUsecase.ItemInput> items = new ArrayList<>();
        for (var dto : req.market_prices()) {
            Optional<StockSymbol> sym = StockSymbol.fromString(dto.symbol());
            if (sym.isEmpty()) {
                CompletableFuture<Void> failed = new CompletableFuture<>();
                failed.completeExceptionally(new BadRequestException("unknown symbol: " + dto.symbol()));
                return failed;
            }
            BigDecimal price;
            try {
                price = new BigDecimal(dto.market_price());
            } catch (RuntimeException e) {
                CompletableFuture<Void> failed = new CompletableFuture<>();
                failed.completeExceptionally(new BadRequestException("invalid market_price: " + dto.market_price()));
                return failed;
            }
            items.add(new UpdateMarketPriceUsecase.ItemInput(sym.get(), price));
        }
        return updateMarketPriceUsecase.run(new UpdateMarketPriceUsecase.Input(items));
    }
}

File: php/src/CodingInterview/Presentation/MarketPriceController.php
<?php

declare(strict_types=1);

namespace Folio\CodingInterview\Presentation;

use Folio\CodingInterview\Application\Usecase\MarketPrice\UpdateMarketPriceItemInput;
use Folio\CodingInterview\Application\Usecase\MarketPrice\UpdateMarketPriceUsecase;
use Folio\CodingInterview\Application\Usecase\MarketPrice\UpdateMarketPriceUsecaseInput;
use Folio\CodingInterview\Domain\BigDecimal;
use Folio\CodingInterview\Domain\StockSymbol;

final class MarketPriceItemDto
{
    public function __construct(
        public readonly string $symbol,
        public readonly string $market_price,
    ) {}
}

final class UpdateMarketPriceRequest
{
    /** @param MarketPriceItemDto[] $market_prices */
    public function __construct(public readonly array $market_prices) {}
}

final class MarketPriceController
{
    public function __construct(private readonly UpdateMarketPriceUsecase $updateMarketPriceUsecase) {}

    public function updateMarketPrice(UpdateMarketPriceRequest $req): void
    {
        $items = [];
        foreach ($req->market_prices as $dto) {
            $sym = StockSymbol::fromStringOrNull($dto->symbol);
            if ($sym === null) {
                throw new BadRequestException("unknown symbol: {$dto->symbol}");
            }
            try {
                $price = new BigDecimal($dto->market_price);
            } catch (\Throwable $e) {
                throw new BadRequestException("invalid market_price: {$dto->market_price}");
            }
            $items[] = new UpdateMarketPriceItemInput($sym, $price);
        }
        $this->updateMarketPriceUsecase->run(new UpdateMarketPriceUsecaseInput($items));
    }
}
